### ISSUE TITLE
Leakage issues

### ISSUE THREAD

**Issue by dugn2119** [aicb-ZhangLabs/Impeller#4]
def prepare_data(data_platform: str = "10XVisium", sample_number: int = 151507, split_number: int = 0, \
                    preprocessed: bool = True, impute_gene_type: str = "all", \
                    use_spatial_graph: bool = False, spatial_graph_type: str = "radius", \
                    spatial_graph_radius_cutoff: int = 150, spatial_graph_knn_cutoff: int = 10, \
                    use_gene_graph: bool = False, gene_graph_knn_cutoff: int = 10, gene_graph_num_high_var_genes: int = 100, \
                    use_heterogeneous_graph: bool = False, \
                    ):

    if data_platform == "10XVisium":
        # At least one type of graph must be created
        assert (use_spatial_graph or use_gene_graph) == True
        if use_heterogeneous_graph:
            assert (use_spatial_graph and use_gene_graph) == True
        # Specify the data directory
        data_dir = "Impeller-main/data/10XVisium/DLPFC"
        if preprocessed:
            adata = sc.read_h5ad(data_dir+"/Preprocessed/"+str(sample_number)+"/filtered_adata.h5ad")
            # Because we need to use id_cell_trans and cell_to_index in the spatial/gene graph
            spatial_graph_coor = pd.DataFrame(adata.obsm['spatial'])
            spatial_graph_coor.index = adata.obs.index
            spatial_graph_coor.columns = ['imagerow', 'imagecol']
            id_cell_trans = dict(zip(range(spatial_graph_coor.shape[0]), np.array(spatial_graph_coor.index)))
            cell_to_index = {id_cell_trans[idx]: idx for idx in id_cell_trans}
            # For spatial graph
            if use_spatial_graph:
                if spatial_graph_type == "radius":
                    nbrs = NearestNeighbors(radius=spatial_graph_radius_cutoff).fit(spatial_graph_coor)
                    distances, indices = nbrs.radius_neighbors(spatial_graph_coor, return_distance=True)
                elif spatial_graph_type == "knn":
                    nbrs = NearestNeighbors(n_neighbors=spatial_graph_knn_cutoff).fit(spatial_graph_coor)
                    distances, indices = nbrs.kneighbors(spatial_graph_coor, return_distance=True)
                else:
                    raise NotImplementedError
                spatial_graph_KNN_list = [pd.DataFrame(zip([it]*indices[it].shape[0], indices[it], distances[it])) for it in range(indices.shape[0])]
                spatial_graph_KNN_df = pd.concat(spatial_graph_KNN_list)
                spatial_graph_KNN_df.columns = ['Cell1', 'Cell2', 'Distance']
                Spatial_Net = spatial_graph_KNN_df.copy()
                Spatial_Net = Spatial_Net[Spatial_Net['Distance']>0]
                Spatial_Net['Cell1'] = Spatial_Net['Cell1'].map(id_cell_trans)
                Spatial_Net['Cell2'] = Spatial_Net['Cell2'].map(id_cell_trans)
                print(f'The spatial graph contains {Spatial_Net.shape[0]} edges, {adata.n_obs} cells.')
                print(f'{Spatial_Net.shape[0]/adata.n_obs:.4f} neighbors per cell on average.')
                adata.uns['Spatial_Net'] = Spatial_Net
                spatial_graph_edge_index = []
                for idx, row in Spatial_Net.iterrows():
                    cell1 = cell_to_index[row['Cell1']]
                    cell2 = cell_to_index[row['Cell2']]
                    spatial_graph_edge_index.append([cell1, cell2])
                    spatial_graph_edge_index.append([cell2, cell1])  # Add reversed edge since the graph is undirected
                spatial_graph_edge_index = torch.tensor(spatial_graph_edge_index, dtype=torch.long).t().contiguous()
            # For gene graph
            if use_gene_graph:
                sc.pp.highly_variable_genes(adata, flavor="seurat_v3", n_top_genes=gene_graph_num_high_var_genes)
                hvg_indices = adata.var['highly_variable']
                hvg_data = torch.from_numpy(adata.X.todense().astype(np.float32))[:, hvg_indices]
                nbrs_gene = NearestNeighbors(n_neighbors=gene_graph_knn_cutoff).fit(hvg_data)
                distances, indices = nbrs_gene.kneighbors(hvg_data, return_distance=True)
                gene_graph_KNN_list = [pd.DataFrame(zip([it]*indices[it].shape[0], indices[it], distances[it])) for it in range(indices.shape[0])]
                gene_graph_KNN_df = pd.concat(gene_graph_KNN_list)
                gene_graph_KNN_df.columns = ['Cell1', 'Cell2', 'Distance']
                Gene_Net = gene_graph_KNN_df.copy()
                Gene_Net['Cell1'] = Gene_Net['Cell1'].map(id_cell_trans)
                Gene_Net['Cell2'] = Gene_Net['Cell2'].map(id_cell_trans)
                print(f'The gene graph contains {Gene_Net.shape[0]} edges, {adata.n_obs} cells.')
                print(f'{Gene_Net.shape[0]/adata.n_obs:.4f} neighbors per cell on average.')
                gene_graph_edge_index = []
                for idx, row in Gene_Net.iterrows():
                    cell1 = cell_to_index[row['Cell1']]
                    cell2 = cell_to_index[row['Cell2']]
                    gene_graph_edge_index.append([cell1, cell2])
                    gene_graph_edge_index.append([cell2, cell1])  # Add reversed edge since the graph is undirected
                gene_graph_edge_index = torch.tensor(gene_graph_edge_index, dtype=torch.long).t().contiguous()
              
            # Create a PyTorch tensor for node features
            x = torch.from_numpy(adata.X.todense().astype(np.float32))
            original_x = x.clone()  # Save the original features

            val_mask = np.load(data_dir+"/Preprocessed/"+str(sample_number)+"/split_"+str(split_number)+"_val_mask.npz")['arr_0']
            test_mask = np.load(data_dir+"/Preprocessed/"+str(sample_number)+"/split_"+str(split_number)+"_test_mask.npz")['arr_0']
            val_mask = torch.tensor(val_mask) 
            test_mask = torch.tensor(test_mask)
            x[val_mask] = 0
            x[test_mask] = 0
            
            if use_spatial_graph and not use_gene_graph:
                data = Data(x=x, edge_index=spatial_graph_edge_index)
            elif not use_spatial_graph and use_gene_graph:
                data = Data(x=x, edge_index=gene_graph_edge_index)
            elif use_spatial_graph and use_gene_graph:
                if use_heterogeneous_graph:
                    edge_index = torch.cat([spatial_graph_edge_index, gene_graph_edge_index], dim=1)
                    edge_type = torch.cat([torch.zeros(spatial_graph_edge_index.size(1), dtype=torch.long),
                                            torch.ones(gene_graph_edge_index.size(1), dtype=torch.long)], dim=0)
                    data = Data(x=x, edge_index=edge_index, edge_type=edge_type)
                else:
                    edge_index = torch.cat([spatial_graph_edge_index, gene_graph_edge_index], dim=1)
                    data = Data(x=x, edge_index=edge_index)
                    
            return data, val_mask, test_mask, x, original_x
My question is
Is it unreasonable to use raw unmasked data when constructing the knn map of gene expression here?

**Comment by Yoyiming** [comment]
I also think there is a leakage issue. @aicb-ZhangLabs
